### PR TITLE
GSoC 2018 Project: PDF generator code for Patient Report section

### PR DESCRIPTION
**Pull request by 2208Abhinav**
This code will create the PDFs for the **Patient Report** section which will be tested using **Test Download** button.
screenshot from 2018-06-04 11-18-28

### CODE REVIEW

**Comment by 2208Abhinav** [comment]
An error for **Demographics** section. Working on it.

**Comment by aethelwulffe** [comment]
Need any help?
-You can always post errors here.

**Comment by aethelwulffe** [comment]
Sorry, I have not tested this yet myself.  I have been working on the PR's from bottom up.

**Comment by aethelwulffe** [comment]
Is this ready to test?

**Comment by 2208Abhinav** [comment]
yes @aethelwulffe its ready to test, just **Demographics** option is having some problems.

File: modules/pdf_generator/report/procedures.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        generate_result_row_pdf($ctx_pdf, $final[0], $rrow_pdf, $final[2]);
+      }
+    }
+  }
+  $content_pro .= "</table><br><br>";
+}
+
+if($i != 1)
+{
+  $pdf->WriteHTML($content_pro, true, false, false, false, '');
```
Comment (aethelwulffe) [inline modules/pdf_generator/report/procedures.php]: Where do I find that function?

**Comment by aethelwulffe** [comment]
What are the dems issues?  Not finding them.  Was this procedures report inclusion the fix?

**Comment by muarachmann** [comment]
@2208Abhinav just looked at your code cool. Went throug  the code and saw the dowload button didnt have a class of cp-output just make sure you dont forget to add that we are going to do a revamping of those stuffs after gsoc

**Comment by aethelwulffe** [comment]
Mua, if you do a "looking at cool code" then WRITE A REVIEW.  K?  We need you here, you are valuable, but we also have you on the hook for this GSOC thing, and we defined community involvement as "writing reviews" among other things.  😛

File: modules/pdf_generator/report/procedures.php (hunk 1 above)
Comment (2208Abhinav) [reply modules/pdf_generator/report/procedures.php]: If you are talking about WriteHTML method then its TCPDF's default method and working on the object $pdf declared in selector.php. Is the code working fine?

**Comment by 2208Abhinav** [comment]
@aethelwulffe the error correction is not included yet in the code. Is the **Procedures Order** code working fine?

**Comment by 2208Abhinav** [comment]
I think for the error that there are some HTML tags not closed properly that is why data is leaking. I will solve it.

**Comment by aethelwulffe** [comment]
Oh...We have always had HTML errors in this junk.  I tend to overlook such things, as none of this has ever worked right.  I have always had...low expectations with this part of the application.  I will adjust the bar up a notch or two!

File: modules/pdf_generator/report/procedures.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    "u.lname AS ulname, u.fname AS ufname, u.mname AS umname " .
+    "FROM procedure_order AS po " .
+    "LEFT JOIN patient_data AS pd ON pd.pid = po.patient_id " .
+    "LEFT JOIN procedure_providers AS pp ON pp.ppid = po.lab_id " .
+    "LEFT JOIN users AS u ON u.id = po.provider_id " .
+    "LEFT JOIN form_encounter AS fe ON fe.pid = po.patient_id AND fe.encounter = po.encounter_id " .
+    "WHERE po.procedure_order_id = ?",
+    $orderid);
+
+    $content_pro .= "(" . oeFormatShortDate(substr($orow['date'], 0, 10)) . ")<br><table border=\"1\" style=\"background-color: silver\"><tr><td>Patient ID</td>".
```
[the remark this answers is not in the document]
Comment (2208Abhinav) [reply modules/pdf_generator/report/procedures.php]: @tmccormi should I use xlt() only with strings like xlt("Procedures Order") or also with the variables like xlt($facility['name']) ?

[the remark this answers is not in the document]
Comment (aethelwulffe) [reply modules/pdf_generator/report/procedures.php]: Only for labels/instructions that are hard-coded.  Never for variables.  The translation is for the application, not any of the contents.  The translations come from an old-school table lookup, so any new text added to applications must also be added to the translations table.

**Comment by aethelwulffe** [comment]
@muarachmann You can filter issues/PR's by choosing the tag "GSOC 2018"

**Comment by 2208Abhinav** [comment]
I think @aethelwulffe is right but still I changed the title @muarachmann 😄

**Comment by 2208Abhinav** [comment]
@aethelwulffe test the code its ready 🤓 but just leave the demographics for now, it will be corrected in next submission.

**Comment by 2208Abhinav** [comment]
@tmccormi due to merge conflict I removed the second download button, but still you can test by using previous single download button.

File: modules/pdf_generator/report/procedures.php [hunk 3]
```diff
@@ -47,35 +56,35 @@ function procedure_order_data($orderid){
...
                     "<td>" . oeFormatShortDate(substr($orow['date'], 0, 10)) . "</td></tr>";
 
-    $content_pro .= "<tr><td>Lab</td>".
+    $content_pro .= "<tr><td>" . xlt("Lab") . "</td>".
                     "<td>" . $orow['labname'] . "</td>";
-    if($orow['specimen_type'])  $content_pro .= "<td>Specimen Type</td>";
-                    
+    if($orow['specimen_type'])  $content_pro .= "<td>" . xlt("Specimen Type") . "</td>";
+
     $content_pro .= "<td>" . $orow['specimen_type'] . "</td></tr></table><br><br>";
```
Comment (muarachmann) [inline modules/pdf_generator/report/procedures.php]: Good that you included the xlt functions but these green lines and red stuffs makes my eyes go :roll_eyes:  will test and get back to a line id needed

**Comment by muarachmann** [comment]
Good you included the xlt functions those green and red lines make my eyes :eyes:  will test and get back to the line in case.

**Comment by muarachmann** [comment]
Worry much about merge conflict for your GSoC project at the end it would be resolved

**Comment by 2208Abhinav** [comment]
@teryhill did you check the code? I want the output PDF of the options selected collectively that I mentioned on the chat.

**Comment by aethelwulffe** [comment]
Checks worked fine, but need to address the zero-date concern from Tony.

**Comment by 2208Abhinav** [comment]
@aethelwulffe I will do it in the end first I am making the whole base then minor issues will be solved.
When I will complete the report section I will do that.

**Comment by aethelwulffe** [comment]
@2208Abhinav OH, I know all that.  Just documenting the status of a test...for the at-the-moment.  I have been really doing poorly keeping up with testing of branches for the last two weeks, and I am just trying to get on top of things.
  For the day-to-day operations of a clinic, retro wrong-headed ideas like exporting things to a PDF document instead of using a data exchange API are actually *vital* to how they do business in the here and now.  The stuff you are working on here directly affects the average user's performance and basic quality of life (less frustration) in a major daily way.  I should be keeping up with testing to help out and verify.

**Comment by 2208Abhinav** [comment]
@tmccormi @teryhill @aethelwulffe I pushed the some new code, removed error for **Demographics** please test the code.

**Comment by aethelwulffe** [comment]
Testing will commence at 15:00 EST

File: modules/pdf_generator/report/insurance.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+    if($data_array%2 != 0)  $content_insu .= "</tr>";
+}
+            
+function insuranceData2 ($retar, $key, $date_format, $content_insu, $i) {
+    global $content_insu;
+    if (@array_key_exists($key,$retar)) {
+        $length = sizeof($retar{$key});
+        if ($retar{$key}[$length]{"value"} != "0000-00-00 00:00:00") {
```
[the remark this answers is not in the document]
Comment (2208Abhinav) [reply modules/pdf_generator/report/insurance.php]: @tmccormi So am I required to use condition like `if(($retar{$key}[$length]{"value"} != "0000-00-00 00:00:00")&&($retar{$key}[$length]{"value"} != NULL))`?

**Comment by aethelwulffe** [comment]
Hi folks,
Torn shoulder keeping me off the computer a lot right now.  
-Testing the Who is Online branch piled on my error log to the point everything shut down.  This branch does it too.
About 100,000 error sets per second are generated repeating this:
```
[Tue Jul 31 13:09:57.485137 2018] [:error] [pid 1068:tid 1956] [client ::1:50491] PHP Notice:  Undefined offset: 1 in C:\\xampp\\htdocs\\branchtest\\library\\appointments.inc.php on line 174, referer: http://localhost/branchtest/modules/calendar/index.php?&framewidth=953
[Tue Jul 31 13:09:57.485137 2018] [:error] [pid 1068:tid 1956] [client ::1:50491] PHP Notice:  Undefined offset: 2 in C:\\xampp\\htdocs\\branchtest\\library\\appointments.inc.php on line 174, referer: http://localhost/branchtest/modules/calendar/index.php?&framewidth=953
[Tue Jul 31 13:09:57.485137 2018] [:error] [pid 1068:tid 1956] [client ::1:50491] PHP Notice:  Only variable references should be returned by reference in C:\\xampp\\htdocs\\branchtest\\library\\encounter_events.inc.php on line 403, referer: http://localhost/branchtest/modules/calendar/index.php?&framewidth=953

```
Sure, those are warnings, but why is it getting called so much?  Basically this seems to be killing my system.  I dare not push the current branch to production.

**Comment by aethelwulffe** [comment]
The errors I am getting are not related to this.
The demographics section looks great....BUUUT it should not (meaning our dems default content) have the social security number....
-Nothing to do with this though.

**Comment by 2208Abhinav** [comment]
@teryhill @aethelwulffe @tmccormi  @nileshprasad137 I did the final commit please check it. I will integrate the PDF generation code for @Trodrige project after GSOC. I have writen the code for it and I am waiting for its integration in main EHR software.

**Comment by 2208Abhinav** [comment]
PDF code for **Review of Systems** and **Review of Systems Checks** is left due due to previous bugs.

**Comment by 2208Abhinav** [comment]
NOTE : The code is left for **Review of Systems** and **Review of Systems Checks**.
And integration with @Trodrige  project is left.

**Comment by aethelwulffe** [comment]
Tested as-is

**Comment by aethelwulffe** [comment]
Review of Systems and Review of Systems Checks is due for replacement itself.  Long overdue.  There are contrib repo chunks related to a possible replacement.

**Comment by aethelwulffe** [comment]
@2208Abhinav Could you please address @tmccormi request to change the comment headers so we can merge this?  It is an awesome and much needed thing.  Pretty please?

**Comment by 2208Abhinav** [comment]
@aethelwulffe I added the new comment headers like it was asked long ago.

**Comment by 2208Abhinav** [comment]
@aethelwulffe what other changes do I need to add for this to be mergeable

**Comment by aethelwulffe** [comment]
I was looking at an old commit.
Yes, this can be merged.
I will do the testing of the PDF's immediately.  If someone wants to review this again, they may do so, but I thing 4-5 months of just sitting here is plenty long enough, and unless they post protests, I am going to merge this pending on my test results.

**Comment by aethelwulffe** [comment]
OK.
Pulled down a pretty extensive PDF which you can find to look at at https://suncoastconnection.com/downloads/screenies/test_rat.pdf.
  If anything, there is only some possible translation or date transformation issues in the code, and even those might not have any real reason to be addressed in this commit.
Merging this and moving on before the NEXT YEAR GSOC starts!
  This fixes a major problem for the EHR.

**Pull request merged**